### PR TITLE
deprecate special "inner constructor" syntax

### PR DESCRIPTION
**Pull request by JeffBezanson**

### CODE REVIEW

**Comment by tkelman** [comment]
`foo{T}(x) where T = rhs` is always going to read to my (and I suspect more than a few others') mental parsers as if it's talking about `T = rhs`. Can we at least add parens to one liner cases of this for single params in base even when not required? `foo{T}(x) where (T) = rhs` at least looks like you might be finished talking about T when you get to the =

**Comment by JeffBezanson** [comment]
There are a few syntax proposals out there. I think some options are

`f(x::T) where {T} = ...`
`f(x::T){T} = ...`
`f(x::T) where T<:Any = ...`  (making `<:Any` either required or strongly encouraged)

**Comment by martinholters** [comment]
Is there a nice way to write code that does not hit a deprecation in 0.6, but still works with 0.5 and possibly even 0.4?
The following works on 0.5 and does not produce a deprecation warning:
```julia
type Foo{T}
    x::T
    (::Type{Foo{T}}){T}(x::T) = new{T}(x)
end
```
It does not work on 0.4, though. It parses, however, so is it worth having
```julia
type Foo{T}
    x::T
    @compat (::Type{Foo{T}}){T}(x::T) = new{T}(x)
end
```
(or another macro in Compat) do the right thing? Obviously, supporting the 0.6 syntax would be much better, but as the `where` did not parse until recently, that's probably not an option.

**Comment by tkelman** [comment]
`@compat` should already be able to translate 0.5-style call overloading in a way 0.4 can handle, did you try it and have problems in some contexts? Maybe since new is a bit special inside type definition blocks that might be handled differently.

**Comment by martinholters** [comment]
Ah, of course, Compat already does this. I less convoluted way of doing this would be nice, but looks like this is indeed usable. Sorry for the noise, then.

**Comment by KristofferC** [comment]
What's the drawback of `f(x::T){T}` compared to `f(x::T) where {T}`. Seems nice to avoid the `where`.

**Comment by JeffBezanson** [comment]
That does seem pretty good for short-form definitions. ` where ` takes up significant space in a 1-liner.

**Comment by yuyichao** [comment]
Not sure if anyone have pointed it out but `f(x::T){T} = ...` parses on 0.4 and only error in lowering so it can be handled in packages / Compat with macros.

**Comment by JeffBezanson** [comment]
Another related issue:

```
function f(x::T)::T where T
```

currently parses as `f(x::T)::(T where T)`, since it is useful to be able to write `x::Array{T} where T` in other contexts, e.g. arguments. (cc @andyferris ) We could change this parsing after a `function` keyword, but I don't think it can be done for short-form definitions. Short-form might look like

```
f(x::T){T}::T = ...
```

which is pretty weird. I think we'd have to allow

```
f(x::T)::T {T}
```

but that introduces a space sensitivity.

**Comment by mbauman** [comment]
It doesn't seem unreasonable to discourage (or maybe even disallow) method type assertions within short form definitions.  If you want to use a method type assertion, you can use the long form… where that special parsing does seem to make sense.

**Comment by andyferris** [comment]
The `where` belongs to the tuple. For example, imagine an unnamed function (where I add a comma to the argument tuple
```julia
function (x::T,) where T
    ...
end
``` 
There has always been an analogy between tuples and function arguments, and the `where` is there to discuss the tuple/argument types, not the output. I would argue the only way add a method type assertion to the above is like this:
```julia
function ((x::T,) where T)::T
    ...
end
```
(you could wrap the `function` in brackets too, not sure if that is worse) or perhaps for named function something like
```julia
function (f(x::T) where T)::T
    ...
end
```
So IMO Jeff's first example should be
```julia
function (f(x::T) where T)::T = ...
```
Would that parse?

**Comment by andyferris** [comment]
I've been thinking - another way to perhaps make this more obvious, as well as the case `f(x::T) where T = T`, would be to replace the `=` with `->` for short form functions.

I'm sure this was brought up once before, but I can't find it right now. Consider the four ways of defining named/anoymous short/long functions in Julia:
```julia
f(x,y) = x + y
(x,y) -> x + y
function f(x,y); x + y; end
function (x,y); x + y; end
```
It is the first one which is the odd-one out (and we can't put `=` in the second). We could have
```
f(x::T) where T -> T
```
which seems more like a function than some `T = T` expression. Using something other than `=` might make it easier to parse the same with method output type assertions syntax, though I'm not sure.

**Comment by StefanKarpinski** [comment]
The snippet `f(x::T) where T -> T` looks to me like it means `f(x::T) where (T -> T)`, which is nonsense, of course. The idea of requiring long form function syntax for return type assertions seems ok to me since there's not really much point in short form definitions with return types.

**Comment by andyferris** [comment]
> looks to me like it means `f(x::T) where (T -> T)`, which is nonsense, of course. 

True. I guess we have the same problem with the current syntax looking like `f(x::T) where (T = T)`. Personally, I think I'll be using `f(x::T where T) = T`, to avoid this.

**Comment by JeffBezanson** [comment]
> `f(x::T where T) = T`

That doesn't mean the same thing.

> `(f(x::T) where T)::T`

I find this too weird --- the last `T` refers to a `T` that is introduced *inside* its expression. To me that just goes against too much intuition about how scope works.

**Comment by andyferris** [comment]
> That doesn't mean the same thing.

Oh right! Oops, I assumed the `T` variable would be visible outside... so apart from allowing triangular dispatch, this is the difference between having the `where` inside and outside the brackets... @JeffBezanson this kind of makes what I was asking for earlier a bit less useful, but I can understand why this is so.

Carry on...

**Comment by andyferris** [comment]
> the last `T` refers to a `T` that is introduced inside its expression.

To me, it seemed similar to `(f(x::T) where T) = ...` where `T` is in scope in the `...`. Variables introduced inside those brackets are visible outside the brackets, at least in some contexts. 

But yes, that syntax does look weird.

**Comment by quinnj** [comment]
This is somewhat continuing the derailing of the PR, but could we make the return type more similar to other languages and put it before the function name? i.e.
```julia
# long-form
function Void f(x, y)

end

# long-form w/ parameters
function T f(x::T, y::T) where T

end

# short-form
Void f(x,y) = ...

# short-form
T f(::T, y::T) where (T) = ...
```

This would be, of course, orthogonal to whether we keep `where` or not; just an idea of where to put the (optional) return type.

**Comment by JeffBezanson** [comment]
I've updated this to use @vtjnash 's suggestion of writing method parameters in curly braces, e.g. `f(x::T, y::S) where {T,S}`. I think this is better since it takes the exact existing syntax, and just moves it after the arguments plus `where`. Moving them after, plus switching curly braces to parens seems a bit scatterbrained. This is also closer to the idea in #20444; if we adopt that then we can simply say the ` where ` part is optional. Also note the symmetry:

```
typealias P{A<:Real, B<:Real} Pair{A,B}

P = Pair{A, B} where {A<:Real, B<:Real}

f(x::Pair{A, B}) where {A<:Real, B<:Real}

struct P{A<:Real, B<:Real}
    ...
end
```

**Comment by andyferris** [comment]
So now the short function form is a bit less confusing?

```julia
f(::T) where {T} = T
```

Looks a bit better than having `T = T` at the end.

PS full symmetry is this:
```julia
P = Pair{A, B} where {A<:Real, B<:Real}

f(x::Pair{A, B}) where {A<:Real, B<:Real}

struct P{A, B} where {A<:Real, B<:Real}
    ...
end
```
but I do love what you have already, Jeff. (Is there some constraints on struct type params that can be expressed in this form but not the existing?

**Comment by JeffBezanson** [comment]
> Is there some constraints on struct type params that can be expressed in this form but not the existing?

I don't believe so. What it would let you express is type parameters that are fixed at a certain value, like

```
struct P{A, B, Int} where {A, B}
```

i.e. the third parameter is always `Int`. This seems pretty useless; I'd rather have the extra conciseness in type definitions. It would probably also break lots of stuff to remove the invariant that the # of parameters of a type equals the # of `where`s.

**Comment by andyferris** [comment]
OK, thanks.

**Comment by o314** [comment]
+1 with andy for the full symmetry syntax.

Thanks you for all the work already done.

File: base/base.jl [hunk 1]
```diff
@@ -152,6 +152,6 @@ immutable Nullable{T}
     hasvalue::Bool
     value::T
 
-    Nullable() = new(false)
-    Nullable(value::T, hasvalue::Bool=true) = new(hasvalue, value)
+    Nullable{T}() where T = new(false)
+    Nullable{T}(value::T, hasvalue::Bool=true) where T = new(hasvalue, value)
```
Comment (tkelman) [inline base/base.jl]: some form of brackets here, please

**Comment by tkelman** [comment]
The readability request, made multiple times by multiple people, was not to have `where T = rhs` for the single parameter case.

**Comment by JeffBezanson** [comment]
We will all be happier if we just get used to it.

**Comment by JeffBezanson** [comment]
Ok, I will add braces to most of these.

**Comment by tkelman** [comment]
There's no line number in the deprecation warning for this, see for example the one from `using Documenter`

**Pull request merged**